Simple webchat for server

tobi opened the pull request:
I put together a simple web-chat that demonstrates how to use the SSE(ish) streaming in the server example. I also went ahead and served it from the root url, to make the server a bit more approachable. 

I tried to match the spirit of llama.cpp and used minimalistic js dependencies and went with the ozempic css style of ggml.ai. 

Initially I went for no-js dependencies but gave up and used a few minimal that i'm importing from js cdns instead of adding them here. Let me know if you agree with this approach. I needed microsoft's fetch-event-source for using event-source over POST (super disappointed that browsers don't support that, actually) and preact+htm for keeping my sanity with all this state,. The upshot is that everything is in one small html file. Speaking of- there is probably a better (and less fragile) way to include the server.html in the cpp binary, but it's been 25 years since I worked with cpp tooling.   


(updated screenshot)
<img width="810" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/347/bda45f50-12d0-45b1-8ca5-d232bc968ee7">

tobi committed:
expose simple web interface on root domain

demonstrates how to use the stream option of generate.

tobi committed:
tighter

ggerganov committed:
server : clear trailing whitespace

ggerganov reviewed (approved):
Love this!

> Initially I went for no-js dependencies but gave up and used a few minimal that i'm importing from js cdns instead of adding them here. Let me know if you agree with this approach. 

I think it's good

> Speaking of- there is probably a better (and less fragile) way to include the server.html in the cpp binary, but it's been 25 years since I worked with cpp tooling.

I guess it would be useful to specify the HTTP root path from the command line arguments instead of hard coding the path. But we can fix this later.

Approving and letting the "server team" take a look and merge

slaren commented:
I think this is a good idea, but the html file should be in the binary. This will not work with the automatic builds because they don't include the contents of the examples directory.

IgnacioFDM commented:
IMHO having the js locally would be better, so it works without an internet connection.

SlyEcho commented:
I have done something like this before, I was serving HTML files in a C HTTP server. There was a CMake option to either build them in or to read them from disk. Reading from files is useful for development because you don't need to rebuild and restart the server. But building them in requires creating a small program that can hexdump the file into a C array definition. Overall pretty complex and then we have the Makefile as well...

> using event-source over POST (super disappointed that browsers don't support that, actually) 

Maybe we could add an endpoint with GET and query parameters?

SlyEcho commented:
> should be pretty simple, i don't touch Makefiles directly often, how bad are custom target?

I'd say Makefiles are a lot easier for this than CMake but it's just added complexity.

@tobi, How hard would it be for you to jam the HTML file contents into the .cpp file?

Green-Sky commented:
after some thinking, i realized that we have pure text file(s), which means we only need to pre and post-fix the file with [raw string literal markers](https://en.cppreference.com/w/cpp/language/string_literal)
eg:
`echo "R\"htmlraw(" > html_build.cpp`
`cat index.html >> html_build.cpp`
`echo ")htmlraw\"" >> html_build.cpp`
in server.cpp:
```c++
const char* html_str =
#include "html_build.cpp"
;
```

Green-Sky commented on examples/server/server.html:
```diff
@@ -0,0 +1,282 @@
+<html>
+
+<head>
+  <title>Llama.cpp</title>
```
i think this should be lowercase :)

Green-Sky commented:
Ok, gave it a go (running it) and found an issue. when ever the window looses focus (switching windows), it restarts the current outputting promt. see screencap
![image](https://github.com/ggerganov/llama.cpp/assets/2938071/786903e3-863c-46b3-be14-26888774d7ff)
(i switched a couple of times back and forth)

SlyEcho commented:
#embed is not gonna work because it's too new.

Yes, it will be harder to develop, but you can also run a simple web server like with Python while developing it.

We can improve it later.

SlyEcho commented:
> cmake script

Cool but we also have to support pure Makefile.

Green-Sky commented:
i feel ignored :sweat_smile: 
we are not dealing with binary files here so my https://github.com/ggerganov/llama.cpp/pull/1998#issuecomment-1607300738 solution is simple 3 text file concats. pretty sure it wont get much simpler :)

SlyEcho commented:
> 3 text file concats

Does it work in Windows?

Green-Sky commented:
> > 3 text file concats
> 
> Does it work in Windows?

if you use make on windows, you likely also have some coreutils installed (`echo` and `cat`)

cmake has built in functions for [read/write/append file](https://cmake.org/cmake/help/v3.6/command/file.html) :)

Green-Sky commented:
> For me, the greatest value of this example is that it demonstrates a minimalistic way of how to implement a basic HTML/JS client that communicates with the server using just a browser without having to install `node` or `curl`. How the client is served can be solved in many different ways, depending on the needs of the specific project. I recommend to merge the example as it is and potentially add improvements from `master`

Agree, except we should really not hard code the path to the html. we basically ship the server, and that would look funky.

@tobi would it be too much to ask to implement the `html root` cli parameter for the server executable?
or for the fasttrack, if the hardcoded .html file could not be loaded (`!file.is_open()`) to fall back to the previous html string?

tobi committed:
minor aesthetic fixes

tobi commented:
sure, i'll try to do that tonight.

tobi committed:
embed index and add --path for choosing static dir

tobi commented:
OK so I did basically all of those things. There is now a --path param that you can point to any directory and static files will be served from this. I also added a deps.sh which just bakes the index.html and index.js into .hpp file (as per @Green-Sky's suggestion). So really, you can launch ./server from the llama.cpp folder and it will use ./examples/server/public directory, copy the ./server file to tmp and it will just use the baked ones, or use --path to work on your own UX. 

The only downside is that we duplicate some files in git here, because of the baked .cpp files. But the deps are so small that it probably doesn't matter. It would be slightly cleaner to go and make deps.sh a build step in cmake and makefile, but... well... I ran out of courage.

tobi committed:
newline police

tobi commented:
@ggerganov server is in reasonably good shape overall. Maybe time for including it in the default build?

IgnacioFDM commented:
I like the current approach, with the website embedded in the binary for simplicity, but also the option to serve from a directory, to improve iteration time and to allow user customization without recompiling. It also includes the js dependencies locally.

I agree with merging this in its current state. Further improvements can be done in future PRs.

🚢

SlyEcho reviewed (changes requested):
Actually, could you move the generated files next to the source files into the `public` folder. If we get more files it will keep it neater to keep to the same directory structure.

ggerganov commented:
> server is in reasonably good shape overall. Maybe time for including it in the default build?

Yes, let's do that. Originally, I insisted to put it behind an option since it was bringing the `boost` library as a dependency, which is a very big burden. Now that the implementation is so self-contained and minimal, we should enable the build by default and maintain it long term

Green-Sky commented on examples/server/index.html.cpp:
```diff
@@ -0,0 +1,278 @@
+// Generated file, run deps.sh to update. Do not edit directly
```
funny that this leading comment work :laughing:

tobi replied:
I thought it was minority clever :-)

Green-Sky commented on examples/server/server.cpp:
```diff
@@ -962,10 +991,9 @@ int main(int argc, char ** argv) {
-        { "hostname", sparams.hostname },
-        { "port", sparams.port },
-    });
+    std::cout << std::endl;
```
1. don't use `std::cout` , we use `(f)printf`
2. don't use `std::endl`, it's excessive. use '\n' instead :)

Green-Sky replied:
@SlyEcho why is the logging mostly json? and inconstant too.

SlyEcho replied:
It should be all JSON. What is inconsistent about it?

SlyEcho replied:
@tobi, please don't change the logging format.

tobi commented:
> Actually, could you move the generated files next to the source files into the `public` folder. If we get more files it will keep it neater to keep to the same directory structure.

that would make it possible to request the files at /index.html.cpp - still want that?

SlyEcho commented:
>> Actually, could you move the generated files next to the source files into the public folder. If we get more files it will keep it neater to keep to the same directory structure.
>
>that would make it possible to request the files at /index.html.cpp - still want that?

I don't think it will because the server only responds to `/` and `/index.js`?

tobi committed:
switched to fprintf logging and to access_log

tobi committed:
allow server to multithread

because web browsers send a lot of garbage requests we want the server
to multithread when serving 404s for favicon's etc. To avoid blowing up
llama we just take a mutex when it's invoked.

tobi commented:
> I don't think it will because the server only responds to `/` and `/index.js`?

it serves the entire directory now because of: 
```cpp
    // Set the base directory for serving static files
    svr.set_base_dir(sparams.public_path);
```

I think that's a good change. And if there is no index.html or index.js it will fall through to the svr.Get request and respond with the baked version. I think it's good to serve the entire sub directory because that makes it super easy for folks to make alternative UX if anyone chooses to. SillyTavern folks might want to give this a go?

developit commented on examples/server/index.html.cpp:
```diff
@@ -0,0 +1,278 @@
+
+      return html`
+        <form onsubmit=${submit}>
+          <input type="text" value="${message}" oninput=${(e) => message.value = e.target.value} autofocus placeholder="Chat here..."/>
+          <button type="submit" disabled=${!generating.value} >Send</button>
+          <button onclick=${(e) => stop(e)} disabled=${generating.value}>Stop</button>
+          <button onclick=${(e) => reset(e)}>Reset</button>
```
```suggestion
          <input value=${message} oninput=${(e) => message.value = e.target.value} autofocus placeholder="Chat here..."/>
          <button type="submit" disabled=${!generating.value} >Send</button>
          <button onclick=${stop} disabled=${generating}>Stop</button>
          <button onclick=${reset}>Reset</button>
```

developit commented on examples/server/index.html.cpp:
```diff
@@ -0,0 +1,278 @@
+
+      useEffect(() => {
+        // scroll to bottom (if needed)
+        if (container.current && container.current.scrollHeight <= container.current.scrollTop + container.current.offsetHeight + 100) {
+          container.current.scrollTo(0, container.current.scrollHeight)
+        }
```
```suggestion
        if (container.current) {
          container.current.scrollTop = 9e9;
        }
```

developit replied:
or:

```suggestion
        container.current?.lastElementChild?.scrollIntoView(false);
```

developit commented on examples/server/index.html.cpp:
```diff
@@ -0,0 +1,278 @@
+          <section class="chat">
+            <${chatStarted.value ? ChatLog : ConfigForm
+        } />
+          </section >
```
```suggestion
          </section>
```

developit commented on examples/server/index.html.cpp:
```diff
@@ -0,0 +1,278 @@
+            <${MessageInput} />
+          </section>
+
+        </main >
```
```suggestion
        </main>
```

developit commented on examples/server/index.html.cpp:
```diff
@@ -0,0 +1,278 @@
+      }, [messages])
+
+      const chatLine = ([user, msg]) => {
+        return html`<p><strong>${template(user, {})}:</strong> ${template(msg, {})}</p>`
```
If messages are ever received out-of-order, you'll want to add a key (you can just use the msg object itself). If they're always appended, the key has no effect and could be left out.

```suggestion
        return html`<p key=${msg}><strong>${template(user, {})}:</strong> ${template(msg, {})}</p>`
```

tobi committed:
improvements

Green-Sky commented:
> If you pull from master, the ci issues should go away.

*merge from master

SlyEcho commented:
> it serves the entire directory now because of:
> ```c++
>    // Set the base directory for serving static files
>    svr.set_base_dir(sparams.public_path);
>```
> I think that's a good change. And if there is no index.html or index.js it will fall through to the svr.Get request and respond with the baked version.

Alright, I can accept it for now.

tobi committed:
revert log format changes

SlyEcho commented:
Well, it seems like MSVC doesn't like it. Another possibility is to turn it into an array with a Python script or maybe `xxd -i`

Green-Sky commented:
`error C2026: string too big, trailing characters truncated ` awwhh, msvc, why do you have to be like that. <.<

SlyEcho commented:
You know what, maybe we can solve the embedded asset problem later?

Green-Sky commented:
> `error C2026: string too big, trailing characters truncated ` awwhh, msvc, why do you have to be like that. <.< edit: apparently msvc only accepts up to 16380 single-byte characters cry

something is funny, they should go up to 32K, except when they are concatenated...

> You know what, maybe we can solve the embedded asset problem later?

sure. +1 from me

tobi committed:
remove need for @microsoft/fetch-event-source dep (-7kb)

tobi committed:
let's try this with the xxd tool instead and see if msvc is happier with that

tobi committed:
enable server in Makefiles

tobi committed:
add /completion.js file to make it easy to use the server from js

tobi commented:
I removed the @microsoft/fetch-event-source dependency and just rolled the fetch by hand. I expose it as `llamaComplete` on `/completion.js` to mirror the `/completion` endpoint. That should make it super easy to use this from js projects. Basically just point server at a directory with your js project, import /completion.js and use llama. 

That got it all under the size limit imposed by msvc. That felt a bit fragile though, so I tried it with `xxd` as @SlyEcho suggested and... well I should have been doing that from the start because it seems that makes everyone happy. 

I also rebased it on master, so I think this is good to go now.

Green-Sky commented:
> Ok, gave it a go (running it) and found an issue. when ever the window looses focus (switching windows), it restarts the current outputting promt. see screencap ![image](https://user-images.githubusercontent.com/2938071/248793268-786903e3-863c-46b3-be14-26888774d7ff.png) (i switched a couple of times back and forth)

did you fix this? did not check yet.

SlyEcho commented on examples/server/server.cpp:
```diff
@@ -950,18 +996,27 @@ int main(int argc, char ** argv) {
+    svr.set_base_dir(sparams.public_path);
+
+    // to make it ctrl+clickable:
+    fprintf(stdout, "\nllama server listening at http://%s:%d\n\n", sparams.hostname.c_str(), sparams.port);
```
```suggestion
    fprintf(stderr, "\nllama server listening at http://%s:%d\n\n", sparams.hostname.c_str(), sparams.port);
```

This needs to not go out with the JSON logs otherwise they could not be parsed correctly.

I think `stderr` is OK for this because other parts of llama.cpp also log there.

SlyEcho commented:
Alright, it's looking very good now. I do have one minor issue and one bigger one:

---

The minor one is the template, is that I think there should be a separate prompts for the message, but for now I can use `### Instruction` and `### Response` for the names if I want to use Alpaca-style models, which seems to work but is missing the line break that the training set has. But I think there could be another template for the message like: 

```
### Instruction:
{{ user }}

### Response:
{{ bot }}
```

---

The major problem right now seems to be that it is not processing whitespace consistently and this causes the server to reevaluate parts of the last response again.

```sh
./bin/server -m ../models/alpaca-native-7b-q4_0.bin -ngl 35 --verbose | jq --unbuffered 'select(.function == "loadPrompt")'
```

This shows how the prompt is handled for each request. Then I did this chat:

```
### Instruction: Can you tell me something funny?
### Response: Why did the chicken cross the playground? To get to the other slide!
### Instruction: Funny.
### Response: I'm glad you enjoyed it!
### Instruction: Tell me another one.
### Response: Why don't scientists trust atoms? Because they make up everything!
```

And the JSON log:

```json
{
  "timestamp": 1688330899,
  "level": "VERBOSE",
  "function": "loadPrompt",
  "line": 223,
  "message": "prompt ingested",
  "n_past": 0,
  "cached": "",
  "to_eval": " This is a conversation between user and llama, a friendly chatbot.\n\n### Instruction: Can you tell me something funny?\n\n### Response:"
}
{
  "timestamp": 1688330912,
  "level": "VERBOSE",
  "function": "loadPrompt",
  "line": 223,
  "message": "prompt ingested",
  "n_past": 32,
  "cached": " This is a conversation between user and llama, a friendly chatbot.\n\n### Instruction: Can you tell me something funny?\n",
  "to_eval": "### Response: Why did the chicken cross the playground? To get to the other slide!\n### Instruction: Funny.\n\n### Response:"
}
{
  "timestamp": 1688330924,
  "level": "VERBOSE",
  "function": "loadPrompt",
  "line": 223,
  "message": "prompt ingested",
  "n_past": 63,
  "cached": " This is a conversation between user and llama, a friendly chatbot.\n\n### Instruction: Can you tell me something funny?\n### Response: Why did the chicken cross the playground? To get to the other slide!\n### Instruction: Funny.\n",
  "to_eval": "### Response: I'm glad you enjoyed it!\n### Instruction: Tell me another one.\n\n### Response:"
}
```

You can see that it is evaluating the bot's response again when it's not needed.

tobi committed:
slightly nicer css

tobi commented:
> > Ok, gave it a go (running it) and found an issue. when ever the window looses focus (switching windows), it restarts the current outputting promt. see screencap ![image](https://user-images.githubusercontent.com/2938071/248793268-786903e3-863c-46b3-be14-26888774d7ff.png) (i switched a couple of times back and forth)
> 
> did you fix this? did not check yet.

i tried to reproduce it across windows, osx, chrome, and safari and couldn't. 


> The minor one is the template, is that I think there should be a separate prompts for the message,

right, so for instruction prompts the chat interface is the wrong modality. It does work however if you switch prompt to:
```
Below is an instruction that describes a task. Write a response that appropriately completes the request.
 
### Instruction:
{{message}}
 
### Response:```

This will just omit the history and {{user}} and {{char}} prompts. I pass {{message}} from the user into the template unchanged, so this comes in handy here. The UX will render in such a way that it looks like a chat, but its really the completion. I think the cleanest way would be to allow switching between chat and instruct mode. That's what most UX do. 

That being said, really all the models that I'm seeing are at least partially chat trained and even untrained instruction tuned models can figure out how to chat from the basic prompts provided. I think it's the right default and we can add first class support for instruct later?


Regarding the prompt stability- well i'm not sure i'm the best person to debug this issue. Are you sure this is a blocker?

SlyEcho commented:
> for instruction prompts the chat interface is the wrong modality.

I don't think it matters, instruction or chat, currently the format is hardcoded like this:
```js
history: transcript.value.flatMap(([name, msg]) => `${name}: ${msg}`).join("\n")
```

That means that for example `${name}:\n${msg}` is not possible.

> Regarding the prompt stability- well i'm not sure i'm the best person to debug this issue. Are you sure this is a blocker?

It is pretty significant, in the worst case the user has to wait quite a long time for the response than necessary.

All of it is from the prompt generation on the client side, not the server.

I think it is duplicating a space character when it saves the response to the history. In the other examples we trim the string for example to avoid this.

SlyEcho commented on examples/server/server.cpp:
```diff
@@ -2,8 +2,6 @@
 #include "llama.h"
 #include "build-info.h"
 
-// single thread
-#define CPPHTTPLIB_THREAD_POOL_COUNT 1
```
I am wondering how this server can be multithreaded now?

tobi replied:
I added a mutex around the llama operations so that the server can multithread and only invocations of llama serialize. This is important now that the server also serves static assets.

SlyEcho replied:
OK, we need to get multithreaded in the future anyway.

tobi commented:
> > for instruction prompts the chat interface is the wrong modality.
> 
> I don't think it matters, instruction or chat, currently the format is hardcoded like this:
> 
> ```js
> history: transcript.value.flatMap(([name, msg]) => `${name}: ${msg}`).join("\n")
> ```
> 
> That means that for example `${name}:\n${msg}` is not possible.

Like I said, instruction mode is totally possible right now. You can ignore the {{history}} part of the prompt. 

I don't think we need an additional template setting for the chat style. `User: Msg` works well on all llamas. There are more elaborate style likes ChatML and we can support those down the line. But outside of those, it's best to just pick one that works reliable and avoid settings overload.

SlyEcho commented:
My concern is that it works differently from main.cpp and all the other examples. They work by adding prefixes before user's messages and a different one before the response. There is no difference between "instruction mode" or "chat mode". This gives a wider compatibility with more models. So, they work the same if it is `Bob: ` or `\n### Response:\n` or `\n### Assistant:` or something else. The name of the participants does not really matter at all. **Anyway, I don't think this is important right now**, the models all seem to be flexible enough to handle it, so I guess it can be improved in the future.

The second part I mentioned should be fixed, though. Otherwise the performance will suffer a lot. The important part here is the whitespace handling: due to the way the tokenizer works the model can return completions with a space added in the front. You can see what is happening if you dump the `payload` variable. I'll try to see if there is some easy fix.

SlyEcho commented on examples/server/public/index.html:
```diff
@@ -0,0 +1,300 @@
+
+        transcript.value = [...history,["{{bot}}", currentMessage]];
+
+        if (data.stop) {
```
I think this solves most of the whitespace issues and doesn't cause the server to reevaluate too much:

```suggestion
        if (data.stop) {
          transcript.value = [...history,["{{bot}}", currentMessage.trim()]];
```

tobi committed:
rework state management into session, expose historyTemplate to settings

tobi committed:
fix mobile, fix missing prompt cache

tobi commented:
> The second part I mentioned should be fixed, though. Otherwise the performance will suffer a lot. The important part here is the whitespace handling: due to the way the tokenizer works the model can return completions with a space added in the front. You can see what is happening if you dump the `payload` variable. I'll try to see if there is some easy fix.

Really good catch. This is now fixed.

tobi committed:
basic response formatting

Green-Sky commented:
> i tried to reproduce it across windows, osx, chrome, and safari and couldn't.

The new code does not have that issue :tada: 




`orca-mini-3b.ggmlv3.q4_0.bin` performing pretty good:

---

User: Tell me a long story about TeddyBear meeting a LLaMa

llama: Once upon a time, in a small village near the mountains, there lived a little girl named Lily. She was known for her kindness and love towards all living beings. One day, while playing in the forest, she stumbled upon a teddy bear that had lost its way. The teddy bear was very sad and lonely, so Lily decided to take it home with her.

As they spent more time together, Lily realized how much the teddy bear needed a friend like her. They would play games, share stories, and snuggle up together under the stars. Their friendship grew stronger every day, and soon the teddy bear became Lily's best companion.

One day, as Lily was walking through the forest, she heard a strange sound coming from a nearby cave. Curious, she ventured inside and found a group of animals gathered around a fire. Among them was a llamma (a type of sheep) who had been lost and alone for days. Lily immediately knew what to do. She took the lllama back to her house, where she fed him some delicious soup and gave him a warm bed to sleep in. With the help of Lily and her kind heart, the lllama was saved and reunited with his friends. From then on, he was never alone again.

And so, TeddyBear met LLaMa, and their friendship blossomed into a beautiful story that will be remembered forever.

Green-Sky reviewed (approved):
Did not see anything wrong.
_Very_ nice addition to our collection. :)

SlyEcho reviewed (approved):
Yeah, it now looks good to me too.

Green-Sky commented:
@tobi i pushed a ci fix to master, if you could pull and rerun ci?

rain-1 commented:
Is it chat only or is there also a text completion UI?

SlyEcho commented:
It is just chat for now.

YannickFricke commented:
@tobi
Is there a specific reason why you went with POST requests for SSE?

And there is (pretty good) support for SSE in browsers: https://developer.mozilla.org/en-US/docs/Web/API/EventSource#browser_compatibility

Green-Sky commented:
> Warning: When not used over HTTP/2, SSE suffers from a limitation to the maximum number of open connections, which can be specially painful when opening various tabs as the limit is per browser and set to a very low number (6). The issue has been marked as "Won't fix" in [Chrome](https://crbug.com/275955) and [Firefox](https://bugzil.la/906896).

countzero commented:
@tobi I love it! Very functional and a great addition! 

I compiled it successfully on Windows (https://github.com/countzero/windows_llama.cpp) and everything works except one CLI setting: The server example does not support the `--n-predict` option. Is this an oversight / bug or intended?

I expected the model options of `server` to be symmetrical to `main`.

SlyEcho commented:
`n_predict` is an API request parameter. The web chat currently has a hardcoded number.

YannickFricke commented:
> > Warning: When not used over HTTP/2, SSE suffers from a limitation to the maximum number of open connections, which can be specially painful when opening various tabs as the limit is per browser and set to a very low number (6). The issue has been marked as "Won't fix" in [Chrome](https://crbug.com/275955) and [Firefox](https://bugzil.la/906896).

Yeah SSE isnt that good compared to Websockets.

So you suggest to switch over to websockets? But it will be pain to implement them as the current httplib doesnt really support them (you would have to do everything on your own)

And when it's the case that a user has more than 6 concurrently open connections to llama?

Another approach could be, that the EventSource is only open while we stream the completions - so you could completly circumvent this issue :)

rain-1 commented:
https://github.com/ggerganov/llama.cpp/pull/1998/commits/dd1df3f31c2c328ccefd68b0c78608371c764733#diff-045455b121ce797624fc9116aaab984486750bee48f02e246708cb168964ec41

I don't like this. Can we please include this as text not octets. It is obfuscated.

rain-1 commented:
There is a crash if the user continues to enter text and hit return to send it while the model is streaming tokens back.

I think it is due to the lock being released in the /completion handler. It may need to be passed inside the thread that is spun off: `const auto chunked_content_provider = [&](size_t, DataSink & sink) {`

Green-Sky commented:
> I don't like this. Can we please include this as text not octets. It is obfuscated.
> 
> An option is to generate the hpp file from the js plain text during compilation.

you can find more context further up (in this thread)

tobi commented:
we should definitely add the xxd runs to the make file instead and remove them from the repo soon

rain-1 commented:
> we should definitely add the xxd runs to the make file instead and remove them from the repo soon because the octet output will balloon the git history otherwise. I do think it's very important to bake the files into the binary though. Honestly, it's crazy that it took C and Cpp so long to add standard ways of doing this.
> 
> We did this in ASM in the 90s all the time.

another option may be to use `ld -b binary`, whatever you prefer :)

SlyEcho commented:
xxd is not a standard tool, it is part of the vim editor. Issue is with portability. Better would be to have our own hex converter in something like Python or C.

Green-Sky commented:
> xxd is not a standard tool, it is part of the vim editor. Issue is with portability. Better would be to have our own hex converter in something like Python or C.

i think he means https://en.cppreference.com/w/c/preprocessor/embed :)

arch-btw commented on examples/server/public/index.html:
```diff
@@ -0,0 +1,359 @@
+
+            <div>
+              <label for="nPredict">Predictions</label>
+              <input type="range" id="nPredict" min="1" max="2048" step="1" name="n_predict" value="${params.value.n_predict}" oninput=${updateParamsFloat} />
```
Should we make `min="-1"` an option for infinity? Thanks!

arch-btw commented:
Oops sorry @tobi & @Green-Sky I didn't realize it had already been merged.
